keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,1018 @@
+1. As a cluster administrator, I want to enforce irrevocable guardrails that
+   all workloads must adhere to in order to guarantee the safety of my clusters.
+2. As a cluster administrator, I want to deploy a default set of policies to all
+   workloads that may be overridden by the developers if needed.
+
+There are several unique properties that we need to add in order accomplish the
+user stories above.
+1. Deny rules and, therefore, hierarchical enforcement of policy
+2. Semantics for a cluster-scoped policy object that may include
+   namespaces/workloads that have not been created yet.
+3. Backwards compatibility with existing Kubernetes Network Policy API
+
+### Non-Goals
+
+Our mission is to solve the most common use cases that cluster admins have.
+That is, we don't want to solve for every possible policy permutation a user
+can think of. Instead, we want to design an API that addresses 90-95% use cases
+while keeping the mental model easy to understand and use.
+
+Additionally, this proposal is squarely focused on solving the needs of the
+Cluster Administrator. It is not intended to solve:
+1. Error reporting for Network Policy
+2. Kubernetes Node policies
+3. New policy selectors (services, service accounts, etc.)
+4. Tools/CLI to debug/explain NetworkPolicies, cluster-scoped policies and their
+   impact on workloads.
+5. Cluster external traffic access control.
+
+## Proposal
+
+In order to achieve the two primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following two new resources
+under `netpol.networking.k8s.io` API group:
+- ClusterNetworkPolicy
+- ClusterDefaultNetworkPolicy
+
+### ClusterNetworkPolicy resource
+
+A ClusterNetworkPolicy resource will help the administrators set strict
+security rules for the cluster, i.e. a developer CANNOT override these rules
+by creating NetworkPolicies that applies to the same workloads as the
+ClusterNetworkPolicy does.
+
+Unlike the NetworkPolicy resource in which each rule represents an allowed
+traffic, ClusterNetworkPolicy will enable administrators to set `Empower`,
+`Deny` or `Allow` as the action of each rule. ClusterNetworkPolicy rules should
+be read as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the ClusterNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+In terms of precedence, the aggregated `Empower` rules (all ClusterNetworkPolicy
+rules with action `Empower` in the cluster combined) should be evaluated before
+aggregated ClusterNetworkPolicy `Deny` rules, followed by aggregated ClusterNetworkPolicy
+`Allow` rules, followed by NetworkPolicy rules in all Namespaces. As such, the
+`Empower` rules have the highest precedence, which shall only be used to provide
+exceptions to deny rules. The `Empower` rules do not guarantee that the traffic
+will not be dropped: it simply denotes that the packets matching those rules can bypass the
+ClusterNetworkPolicy `Deny` rule evaluation.
```

Comment:
We were discussing this on the sig-network-policy meeting yesterday, and I want to clarify the meaning of `Empower` as it stands a bit more here:
There has been two schools of thought on how `Empower` should be interpreted: 1. `Empower` provides exception to Deny rules of ClusterNetworkPolicy  2. `Empower` delegates policy control of specific traffic to the developers. While 1. is always true,  the delegation happens when there are no ClusterNetworkPolicy Allow rules explicitly dictating that traffic must be allowed (which must be a subset of what's Empowered, in the presence of some sort of Deny-all rule). Consider the following example (assuming each tenant owns single namespace):
 ```
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        type: tenant
  ingress:
    - action: Empower
      from:
        - namespaceSelector:
            matchLabels:
               app: prometheus
        - namespaces:
             scope: Self
    - action: Deny
      from:
        - namespaceSelector: {}
    - action: Allow
      from:
        - namespaceSelector:
             matchLabels:
               app: prometheus   
```
There are 3 different kind of traffic affected: 
1. Prometheus namespace -> tenant namespace: Always allowed, as it is under `Empower` AND `Allow`. This might be a requirement for compliance purposes for example.

2. tenant A -> tenant A, tenant B -> tenant B etc: Delegated to developers of each tenant namespace. Those traffic are not Denied because of `Empower`, but not explicitly allowed either. Namespace owners can use k8s NetworkPolicies to further regulate intra-namespace traffic.

3. tenant B -> tenant A, tenant A -> tenant B etc: Dropped.

Hope it will clarify things a bit more, and feedbacks are also welcomed.

Reply:
resolving as this is regarding an older proposal we had